R/vpc-simulation.R

```diff
@@ -0,0 +1,253 @@
+  # Create new simulation model object
+  model_dir <- get_model_working_directory(.mod)
+  sim_dir <- glue("{get_model_id(.mod)}-{.suffix}")
+
+  .sim_mod <- copy_model_from(
+    .parent_mod = .mod,
+    .new_model = sim_dir,
+    .add_tags = "SIMULATION",
+    .inherit_tags = .inherit_tags,
+    .update_model_file = TRUE,
+    .overwrite = .overwrite
+  )
+
+  .sim_mod[[YAML_MOD_TYPE]] <- "nmsim"
+  .sim_mod <- save_model_yaml(.sim_mod)
+
+  # Change problem statement
+  prob <- glue("Simulation of model {get_model_id(.mod)}")
+  modify_prob_statement(.sim_mod, prob)
+
+  # Set up simulation
+  setup_sim_run(.sim_mod, n = n, seed = seed, replicate_var = replicate_var)
+
+  # Return read-in model to get the updated class as part of the model object
+  return(read_model(file.path(model_dir, sim_dir)))
+}
+
+
+#' Set up `NONMEM` simulation
+#'
+#' Creates new `$SIMULATION` and `$MSFI` records, and modifies the `$PRED` or
+#'  `$ERROR` record (whichever exists) to have the additional line: `REPI=IREP`.
+#' @param .mod a `bbi_nonmem_model` or `bbi_nmsim_model` object.
+#' @inheritParams new_sim_model
+#' @keywords internal
+setup_sim_run <- function(.mod, n = 100, seed = 1234, replicate_var = "REPI"){
+  check_model_object(.mod, c(NM_MOD_CLASS, NMSIM_MOD_CLASS))
+  checkmate::assert_numeric(n, lower = 1)
+  checkmate::assert_numeric(seed)
+
+  # Remove the following record types for simulation:
+  # $EST, $COV,  $TABLE, $SIMULATION
+  records_rm <- c("estimation", "covariance", "table", "simulation")
+  purrr::walk(records_rm, function(rec){
+    if(mod_has_record(.mod, type = rec)) remove_records(.mod, type = rec)
+  })
+
+  # Remove PK and prior records:
+  # $PRIOR, $THETA/$THETAP/$THETAPV, $OMEGA/$OMEGAP/$OMEGAPD, $SIGMA/$SIGMAP/$SIGMAPD
+  records_rm <- c("prior", "theta", "thetapv", "omega", "omegapd", "sigma", "sigmapd")
+  purrr::walk(records_rm, function(rec){
+    if(mod_has_record(.mod, type = rec)) remove_records(.mod, type = rec)
+  })
+
+
+  ## Add new variable assignment to $ERROR _or_ $PRED: `REPI=IREP` ##
+  # Store current $error or $pred record
+  rep_name <- get_sim_replicate_record(.mod)
+  rep_lines <- get_records(.mod, rep_name, get_lines = TRUE)[[1]]
+  rep_lines <- rep_lines[rep_lines != ""]
+  rep_lines <- c(rep_lines, glue("{replicate_var}=IREP"), "")
+
+  # Remove old record, and add a new one with the modification
+  #  - Use existing record name, dont overwrite
+  #  - easiest way to amend an existing record
+  remove_records(.mod, rep_name)
+  add_new_record(.mod, rep_name, rec_name = NULL, lines = rep_lines, after = NULL)
+
+
+  ## Add new $SIMULATION record (must be _after_ $ERROR or $PRED, but before $TABLE) ##
+  sim_lines <- glue("({seed}) ({seed} NORMAL) SUBPROBLEMS={n} TRUE=FINAL")
+  add_new_record(.mod, "simulation", lines = sim_lines, after = rep_name)
+
+
+  ## Add new $TABLE record with predefined format and columns (including REPI) ##
+  table_name <- glue("{get_model_id(get_based_on(.mod)[1])}sim.tab")
```

Comment:
For run simulation run `106-sim`, the control stream is named `106-csim` and  table is named `106sim.tab` Suggest to keep the table name consistent with the run root.

Reply:
Oh I actually did that because I was under the impression `NONMEM` didnt support special characters. Assuming that's not true for file paths ill make the change

Reply:
Confirmed this was fine and made the change (currently uncomitted)